crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -685,7 +685,12 @@ impl<'db> TypeInferenceBuilder<'db> {
                     let symbol = symbols.symbol_by_name(id).unwrap();
                     if !symbol.is_defined() || !self.scope.is_function_like(self.db) {
                         // implicit global
-                        Some(module_global_symbol_ty_by_name(self.db, self.file, id))
+                        let mut unbound_ty =
+                            module_global_symbol_ty_by_name(self.db, self.file, id);
+                        if matches!(unbound_ty, Type::Unbound) {
+                            unbound_ty = builtins_symbol_ty_by_name(self.db, id);
+                        }
```

Comment:
I think one issue shown by the benchmarks might be that we're getting cycles? I think we should ensure that if we are inferring types in the builtins scope, we _don't_ make this fallback.

Reply:
Ah! Very good point.

Reply:
This doesn't actually result in a cycle currently, because querying a public symbol type is different from querying a use type. But it does result in a lot of extra work, which I suspect is the cause of the regression.